File: sentiment_analyzer.py
from typing import List, Dict
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from models import RedditPost, SentimentResult


class SentimentAnalyzer:
    def __init__(self):
        """Initialize the sentiment analyzer with VADER."""
        self.analyzer = SentimentIntensityAnalyzer()
    
    def get_sentiment_score(self, text: str) -> SentimentResult:
        """
        Analyze sentiment of a single text snippet using VADER.
        
        Args:
            text: Text to analyze for sentiment
            
        Returns:
            SentimentResult with compound score and sentiment category
        """
        if not text or not text.strip():
            return SentimentResult(0.0, 0.0, 0.0, 1.0, "Neutral")
        
        # Get VADER sentiment scores
        scores = self.analyzer.polarity_scores(text)
        
        # Determine sentiment category based on compound score
        compound = scores['compound']
        if compound > 0.1:
            category = "Positive"
        elif compound < -0.1:
            category = "Negative"
        else:
            category = "Neutral"
        
        return SentimentResult(
            compound_score=compound,
            positive=scores['pos'],
            negative=scores['neg'],
            neutral=scores['neu'],
            category=category
        )
    
    def analyze_stock_sentiment(self, posts: List[RedditPost], ticker: str) -> SentimentResult:
        """
        Calculate overall sentiment for a specific stock across multiple posts.
        
        Args:
            posts: List of RedditPost objects to analyze
            ticker: Stock ticker symbol to analyze sentiment for
            
        Returns:
            SentimentResult with aggregated sentiment for the stock
        """
        relevant_texts = []
        ticker_upper = ticker.upper()
        
        # Collect all text snippets that mention the ticker
        for post in posts:
            # Check title
            if post.title and ticker_upper in post.title.upper():
                relevant_texts.append(post.title)
            
            # Check content
            if post.content and ticker_upper in post.content.upper():
                relevant_texts.append(post.content)
            
            # Check comments
            for comment in post.comments:
                if comment and ticker_upper in comment.upper():
                    relevant_texts.append(comment)
        
        if not relevant_texts:
            return SentimentResult(0.0, 0.0, 0.0, 1.0, "Neutral")
        
        # Calculate sentiment for each relevant text
        sentiment_scores = []
        positive_sum = 0.0
        negative_sum = 0.0
        neutral_sum = 0.0
        
        for text in relevant_texts:
            sentiment = self.get_sentiment_score(text)
            sentiment_scores.append(sentiment.compound_score)
            positive_sum += sentiment.positive
            negative_sum += sentiment.negative
            neutral_sum += sentiment.neutral
        
        # Calculate average compound score
        avg_compound = sum(sentiment_scores) / len(sentiment_scores)
        
        # Calculate average component scores
        count = len(relevant_texts)
        avg_positive = positive_sum / count
        avg_negative = negative_sum / count
        avg_neutral = neutral_sum / count
        
        # Determine overall category
        if avg_compound > 0.1:
            category = "Positive"
        elif avg_compound < -0.1:
            category = "Negative"
        else:
            category = "Neutral"
        
        return SentimentResult(
            compound_score=avg_compound,
            positive=avg_positive,
            negative=avg_negative,
            neutral=avg_neutral,
            category=category
        )
    
    def analyze_multiple_stocks(self, posts: List[RedditPost], tickers: List[str]) -> Dict[str, SentimentResult]:
        """
        Analyze sentiment for multiple stocks at once.
        
        Args:
            posts: List of RedditPost objects to analyze
            tickers: List of stock ticker symbols to analyze
            
        Returns:
            Dictionary mapping ticker symbols to their sentiment results
        """
        results = {}
        
        for ticker in tickers:
            results[ticker] = self.analyze_stock_sentiment(posts, ticker)
        
        return results
    
    def get_sentiment_summary(self, posts: List[RedditPost]) -> Dict[str, int]:
        """
        Get overall sentiment distribution across all posts.
        
        Args:
            posts: List of RedditPost objects to analyze
            
        Returns:
            Dictionary with counts of positive, negative, and neutral posts
        """
        sentiment_counts = {"Positive": 0, "Negative": 0, "Neutral": 0}
        
        for post in posts:
            # Combine title and content for overall post sentiment
            combined_text = ""
            if post.title:
                combined_text += post.title + " "
            if post.content:
                combined_text += post.content
            
            if combined_text.strip():
                sentiment = self.get_sentiment_score(combined_text)
                sentiment_counts[sentiment.category] += 1
        
        return sentiment_counts

File: reddit-sentiment-next/src/lib/services/sentiment-analyzer.ts
// Port of Python sentiment_analyzer.py to TypeScript using the sentiment library

import Sentiment from 'sentiment';
import { RedditPost, SentimentResult } from '@/types';

export class SentimentAnalyzer {
  private analyzer: Sentiment;

  constructor() {
    this.analyzer = new Sentiment();
  }

  getSentimentScore(text: string): SentimentResult {
    if (!text || !text.trim()) {
      return {
        compound_score: 0.0,
        positive: 0.0,
        negative: 0.0,
        neutral: 1.0,
        category: 'Neutral',
      };
    }

    // Get sentiment analysis result
    const result = this.analyzer.analyze(text);
    
    // Convert score to a normalized compound score similar to VADER
    // The sentiment library returns a score typically between -5 and 5
    // We'll normalize it to -1 to 1 range like VADER
    const compound = Math.max(-1, Math.min(1, result.score / 5));
    
    // Calculate normalized positive/negative/neutral scores
    // This is a simplified approach compared to VADER's more sophisticated scoring
    const absScore = Math.abs(compound);
    let positive = 0.0;
    let negative = 0.0;
    const neutral = 1.0 - absScore;

    if (compound > 0) {
      positive = absScore;
    } else if (compound < 0) {
      negative = absScore;
    }

    // Determine sentiment category based on compound score
    let category: 'Positive' | 'Negative' | 'Neutral';
    if (compound > 0.1) {
      category = 'Positive';
    } else if (compound < -0.1) {
      category = 'Negative';
    } else {
      category = 'Neutral';
    }

    return {
      compound_score: compound,
      positive,
      negative,
      neutral,
      category,
    };
  }

  analyzeStockSentiment(posts: RedditPost[], ticker: string): SentimentResult {
    const relevantTexts: string[] = [];
    const tickerUpper = ticker.toUpperCase();

    // Collect all text snippets that mention the ticker
    for (const post of posts) {
      // Check title
      if (post.title && post.title.toUpperCase().includes(tickerUpper)) {
        relevantTexts.push(post.title);
      }

      // Check content
      if (post.content && post.content.toUpperCase().includes(tickerUpper)) {
        relevantTexts.push(post.content);
      }

      // Check comments
      for (const comment of post.comments) {
        if (comment && comment.toUpperCase().includes(tickerUpper)) {
          relevantTexts.push(comment);
        }
      }
    }

    if (relevantTexts.length === 0) {
      return {
        compound_score: 0.0,
        positive: 0.0,
        negative: 0.0,
        neutral: 1.0,
        category: 'Neutral',
      };
    }

    // Calculate sentiment for each relevant text
    const sentimentScores: number[] = [];
    let positiveSum = 0.0;
    let negativeSum = 0.0;
    let neutralSum = 0.0;

    for (const text of relevantTexts) {
      const sentiment = this.getSentimentScore(text);
      sentimentScores.push(sentiment.compound_score);
      positiveSum += sentiment.positive;
      negativeSum += sentiment.negative;
      neutralSum += sentiment.neutral;
    }

    // Calculate average compound score
    const avgCompound = sentimentScores.reduce((sum, score) => sum + score, 0) / sentimentScores.length;

    // Calculate average component scores
    const count = relevantTexts.length;
    const avgPositive = positiveSum / count;
    const avgNegative = negativeSum / count;
    const avgNeutral = neutralSum / count;

    // Determine overall category
    let category: 'Positive' | 'Negative' | 'Neutral';
    if (avgCompound > 0.1) {
      category = 'Positive';
    } else if (avgCompound < -0.1) {
      category = 'Negative';
    } else {
      category = 'Neutral';
    }

    return {
      compound_score: avgCompound,
      positive: avgPositive,
      negative: avgNegative,
      neutral: avgNeutral,
      category,
    };
  }

  analyzeMultipleStocks(posts: RedditPost[], tickers: string[]): Record<string, SentimentResult> {
    const results: Record<string, SentimentResult> = {};

    for (const ticker of tickers) {
      results[ticker] = this.analyzeStockSentiment(posts, ticker);
    }

    return results;
  }

  getSentimentSummary(posts: RedditPost[]): Record<string, number> {
    const sentimentCounts = { Positive: 0, Negative: 0, Neutral: 0 };

    for (const post of posts) {
      // Combine title and content for overall post sentiment
      let combinedText = '';
      if (post.title) {
        combinedText += post.title + ' ';
      }
      if (post.content) {
        combinedText += post.content;
      }

      if (combinedText.trim()) {
        const sentiment = this.getSentimentScore(combinedText);
        sentimentCounts[sentiment.category] += 1;
      }
    }

    return sentimentCounts;
  }
}